### DISCUSSION TITLE
[RFC] Allow Importing Modules via URLs

### DISCUSSION THREAD

**Discussion by lsh** [modular/modular#413] [Ideas]
# Overview
One of Mojo's main priorities is solving the "two language problem," which means that it must function for both app development use cases, but also one off scripts. [Dependency management is not Python's strong suit](https://pradyunsg.me/blog/2023/01/21/thoughts-on-python-packaging/). Mojo will have to navigate this problem space, and in the fail case, users may have to manage Python dependencies for scripts, and Mojo dependencies for apps. In order to prevent this problem, and promote ergonomic scripting, I propose allowing the following (syntax to be bikeshed):
```python
from "url" import method
```
Allowing importing from URLs has many benefits from security, organizational, and ergonomic standpoints, which we will cover below.

# Out of Scope

This proposal makes little-to-no claims about what _defines_ a module in Mojo. A module could be a binary or source code. The only assumption is that the module contains some form of manifest (e.g. `Cargo.toml`, `pyproject.toml`, `package.json`) that allows for resolution of the module.

Another aspect that is out of scope is _dependency type resolution_. In other words, if two versions of the same package have the same struct with the same ABI, are they the same type? That question will have to be answered, but not in this proposal.

Furthermore, while this project discusses package registries as a concept, there is no suggestions about Mojo having a registry, and any discussion of that is probably best left to a separate proposal.

Finally, this project makes no assumptions about the build system or compilation process itself. Mojo may decide to use a `mojo.toml`, `pyproject.toml`, or even go the Zig route of allowing users to use Mojo code itself to compile their code.

# Background

## Deno
[Deno](https://deno.land/) is the main inspiration behind this proposal. In Deno, users import modules directly from a url or path.
This import method looks like:
```ts
// remote
import {
  add,
  multiply,
} from "https://x.nest.land/ramda@0.27.0/source/index.js";
// local
import { add, multiply } from "./arithmetic.ts";
```
A benefit of this style of imports is that every script comes with a built in list of dependencies, so a user can run `deno run file.ts`, and it will pull in all dependencies from the specified URLs. [Alex Kladov has written](https://matklad.github.io/2023/02/12/a-love-letter-to-deno.html) about how Deno's module system helps remove the need to bundle a script or package inside a docker container or virtual environment.

Using this import method, [a common convention in Deno](https://deno.com/manual@v1.34.3/examples/manage_dependencies) is to have a `deps.ts` file in a project, which would look roughly like:
```ts
// in deps.ts
export * as io from "https://deno.land/std@0.192.0/io/mod.ts";

// in file.ts
import { io } from "./deps.ts";
```

Deno also allows the use of [import maps](https://deno.com/manual@v1.34.3/basics/import_maps) in a [Deno config file](https://deno.com/manual@v1.34.3/getting_started/configuration_file). These mappings allow users to alias file paths and urls.

Finally, and most relevant to Mojo, Deno was faced with a large existing ecosystem of centralized packages in NPM. The way Deno handled this problem was by allowing the import syntax `npm:package`. In other words, users can do:
```ts
import { map } from "npm:lodash";
```

## Go

In Golang, files are imported with the syntax:
```go
import (
  "fmt" // a package that exists in the std library
  "go.opentelemetry.io/otel" // a package hosted by opentelemetry.io
  "github.com/gin-gonic/gin" // a package hosted on GitHub
)
```
Go does not have a central package repository, in contrast to crates.io, npm, or pip. Users download third party modules via `go get module`. Go also uses URLs as module identifiers, which helps for keeping the provenance of installed modules. One consideration is that Go files are not designed to be run as scripts in contrast to Deno and Python.

An important note is that `go.mod` also includes the version of Go used when creating the project.This decision is helpful for reproducible builds. 

## Rust
Rust maintains dependencies in the `Cargo.toml` manifest. This strategy makes sense for a build system, but as Rust begins to consider how it will handle scripting use cases, this strategy becomes harder to manage. Currently, there is [the `cargo-script` pre-rfc](https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639), and the suggested way of handling dependencies in the proposal is a listing in comments:
```rust
#!/usr/bin/env cargo

//! ```cargo
//! [dependencies]
//! clap = { version = "4.2", features = ["derive"] }
//! ```

use clap::Parser;

#[derive(Parser, Debug)]
#[clap(version)]
struct Args {
    #[clap(short, long, help = "Path to config")]
    config: Option<std::path::PathBuf>,
}

fn main() {
    let args = Args::parse();
    println!("{:?}", args);
}
```
While this syntax makes sense for attempting to add scripting to an established systems language such as Rust, it is not as ergonomic as directly importing from a URL like is done in Deno. This example may also serve to illustrate how difficult it is to handle the scripting case later rather than considering it from the start.

## Zig
Zig recently added a package manager. Projects are managed in a `projectName.zon` file (short for Zig Object Notation). When adding a dependency to a project, a `build.zig.zon` might look like (example taken from [getty](https://getty.so/user-guide/installation/)):
```
.{
    .name = "my-project",
    .version = "0.0.0",
    .dependencies = .{
        .getty = .{
            .url = "https://github.com/getty-zig/getty/archive/<COMMIT>.tar.gz",
        },
    },
}
```
Zig also does not have a central package repository, and so users import dependencies directly from urls.

# Proposal

In order to allow easy dependency management in scripts, I propose allowing the ability to import modules directly from URLs. So, if there were a package such as `github.com/modularml/matrix` that contained a `Matrix` struct, users could run:
```python
from "github.com/modularml/matrix" import Matrix
# alternatively
import "github.com/modularml/matrix" as matrix
```
And in the vein as Deno, importing from pip would be formatted as:
```python
import "pip:requests" as requests
```
with the special case that pip importants would automatically be `PythonModules`s.

There are several benefits that come from this syntax right off the bat:
* Modular will allow users to share scripts quickly, which may gain some time for fleshing out the build system.
* Modular will not have to immediately figure out its relationship to pip, nor will it have to make a decision on hosting a package registry any time soon.
* Users will be able to create reproducible scripts.

There is also the security benefit to URL based package management. In Python, when a user calls `from my_package import func`, there is little information about the source of the package. Most users will assume it was installed via pip, though it could be Conda. Furthermore, tracking the exact version installed is a mess, and `__init__.py` files will check to make sure installed package versions are compatible. While one still does have to be careful about what package or URL they import, it is at least easier to figure out the entire contents of the build from this method.

One last benefit to this scheme is that because it is so general, it can also extend to the C/C++ interrop once that becomes available. Currently, the Mojo roadmap describes the FFI import to look like:
```python
from "math.h" import cos
```
Using the URL/path approach, this becomes:
```python
# sys header, bikeshed key
from "header:math.h" import cos
```

# Conclusion
In conclusion, this proposal would allow the Mojo community to begin sharing scripts with each other, while pushing off registry questions. It would also make the provenance of dependencies clear. Furthermore, it would help to make dependencies in Mojo work for scripts, and not just full packages. Finally, it would make the process of sharing scripts easy, since users would not have to package all of their dependencies.

# Unresolved

There are still some unresolved questions about how to handle the local storage of a file after it is resolved. Some type of caching will be important, but the question of how to handle multiple versions of the same package is tough. In Deno, this is handled by encouraging a target triple in the URL.

Another unresolved portion is the mapping of package names from the project manifest to urls. I would expect that for Mojo, it makes sense to have something like a `mojo.toml` (bikeshed on format/name/etc), where a dependency may look like:
```toml
my_package = "example.com/pkg"
```
and could then be imported into a Mojo project with `import my_package`.

**Comment by mojodojodev** [comment]
_edited to match the better suggestion from @cjohnson318 in the replies_

This is a great proposal, if the packaging is built into the source code without a separate `toml`, Mojo would be a major draw card for users that just want to build and run scripts as well.

A couple of extra things to add:

```python
import "pip:requests" as requests
```
Looks great but will need extra functionality for version numbers, how about making it work the same way as the `pip` command line or `requirements.txt`:
```python
import "pip:requests == 2.31.0" as requests
import "pip:requests[security] >= 2.8.1" as requests
import "pip:requests[security]==2.8.*" as requests
```
Which can be shared with the non-central repository version:
```python
import "github.com/modularml/matrix == 0.1.0" as matrix
```

Also in the current Python ecosystem there is quite a few issues with cross platform versioning and packaging, lots of things that run fine on x86-64 linux but don't work on arm64 macos e.g. for tensorflow on m1 you have to add `tensorflow-metal` and `tensorflow-macos`.

Could use the same mechanism that already exists for conditional compilation, although it might need a little magic to hoist for example `tf` to the outer scope on `import` statements:

```rust
from TargetInfo import os_is_macos, is_apple_m1

@parameter
if os_is_macos():
    import "pip:tensorflow-macos" as tf
else:
    import "pip:tensorflow" as tf

@parameter
if is_apple_m1():
    import "pip:tensorflow-metal"
```

**Reply by mojodojodev** [reply]
And also this: `even go the Zig route of allowing users to use Mojo code itself to compile their code` is an incredibly good idea for users that need to build/link to external C/C++ code in their project, `build.rs` in Rust is a whole extra un-intuitive thing you have to learn and I love the way they do this in Zig.

**Reply by cjohnson318** [reply]
I really like the idea of keeping requirements as code, as in the TypeScript example, and not having a separate subsystem for managing requirements as in Python, e.g. pip, poetry, conda, etc.

One suggestion is that instead of introducing a new keyword `tag`, one could add version modifiers into the string like,

```
import "pip:requests" as requests
import "pip:requests == 2.31.0" as requests
import "pip:requests[security] >= 2.8.1" as requests
import "pip:requests[security] == 2.8.*" as requests
import "github.com/modularml/matrix == 0.1.0" as matrix
```

**Reply by mojodojodev** [reply]
@cjohnson318 yeah that looks much better, the pip requirements are then the exact same as when you're using the pip command line or a `requirements.txt`

**Comment by lattner** [comment]
This is a very nice proposal, and I agree there is utility to this.  We need to get more basic compilation model basics nailed down (as you point out) and aren't anywhere near touching into packaging and other related issues that would be a depenency.  I don't see us being able to do much in this area for many months.

**Comment by Mogball** [comment]
Thanks for the proposal! It seems quite sensible. I'm +1 in this direction for having URL imports. Traditionally, it has been a responsibility placed on the build system, but that makes reproducible scripts harder to manage. Having first-class support for URL imports seem like the right direction for Mojo.

We're still focusing on nailing down having packages first, but later on thinking about distribution and sharing will be important to building an ecosystem.

**Comment by ksandvik** [comment]
Imports with URLs sounds nice. My only worry are dependency handling and there's a underlying system to handle specific versions to download. Anyway, this is in future.

**Comment by drunkwcodes** [comment]
What a terrible feature and consequence it brings.

It adds unnecessary network stack and errors to core language function (import), and make a constraint to programs to only run on a healthy network.

URL import is kind of a bad habit to me.

**Reply by lsh** [reply]
I'd like to understand your concerns a little more, or hear any alternative suggestions you have for how to handle dependencies in scripts.

> It adds unnecessary network stack and errors to core language function (import), and make a constraint to programs to only run on a healthy network.

In terms of actual network use, none of the precedents above make a network call every single time. They only check the first time if dependencies are cached.

This is meant to be functionally equivalent to sending a requirements.txt along with a script, and a user installing the dependencies required, then running the script.

Importantly, this also isn't saying this method should be the only technique to import modules, but for users that care about build reproducibility and portability, this feature has many benefits.

**Reply by drunkwcodes** [reply]
Mojo has a simple compiler, so the error handling things may be up to users. We have to write those catching 404 error, timeout limit exceed error etc in our code if we introduce URL import feature into Mojo.

I don't need Zig & Sharko say: hey let's do this.

Sharko: "I DON'T NEED PRE-INSTALLED AND WELL-TESTED PACKAGE AND SOFTWARE, I ONLY NEED FLOATING HACKED CRAP VERSION SCRIPTS ON THE WEB."

Please no.

If introducing URL imports, somebody would not make an official release, upload to central package index, only provide a link like a gibbering bash script, IN MOJO.
It is certainly not a formal way to deliver scripts, not even more formal than making a zip and put it in a USB.

If it's baked in the language, more and more people will find the code base being scattered on the web, depending on Github, depending on mysterious NAS, depending on some CDN don't know when it collapse.

Mojo is good at scaling, not making the code base be a distributed mess. It would be a spoiled, deteriorated, decaying, and rotten language if this thing keeps going on!

Please convince me by telling the burden to do an official release, using git, or installing scripts at first. I still don't get what this feature wants to achieve.

Mojo is not javascript, nor Zig and Sharko.

**Reply by lsh** [reply]
> We have to write those catching 404 error, timeout limit exceed error etc in our code if we introduce URL import feature into Mojo.

I'm not sure how that would be the case since this would only occur at compile time, and only the single time to make sure the dependency is installed.

> If introducing URL imports, somebody would not make a official release, upload to central package index, only provide a link like a gibbering bash script, IN MOJO.

This statement also seems unfounded and is not the case in any of the other languages listed. As for the central package index part, there is no public information right now about how Mojo plans to handle the package ecosystem, or whether they will attempt to host a central package index. Software supply chain security is certainly important, but it becomes weaker when users have a ton of packages installed with no concept of where they came from. Is a given piece of software installed via pip, homebrew, or an installer? Having a source of truth listing where packages came from is useful for answering those kinds of questions.

Packages go through breaking changes, which is why people pin dependencies. This feature addresses similar problems to Docker, Nix, and virtual environments. It makes it clear where every dependency came from, and that they are all present at compile time.

As for your Zig and Sharko comments, I assume this was some attempt at being snarky, but I'm not entirely sure.

**Comment by ksandvik** [comment]
I think the issue is that if there's an URL for network downloading, then the compiler has to have code to download and parse the data as import is a keyword. Well, it could be be done, especially with meta-language programming, but it adds complexity.

**Reply by lsh** [reply]
I'm not completely sure I understand the concern here, so I would appreciate a more thorough explanation.

Mojo will in all likelihood have a build system built in, which will include package management, so there will surely be some form of networking interface already in the codebase. Python also has URL parsing as part of the standard library. Is there something here that I'm missing?

**Comment by w32zhong** [comment]
I don't understand why we need URL when we can specify released package name?

**Comment by richardkiss** [comment]
Importing objects by a human-readable name that does include a cryptographic has leads to vulnerabilities by swapping the back-end package. See for example

https://www.csoonline.com/article/643153/millions-of-github-repositories-vulnerable-to-repojacking-report.html

Any cryptographically secure packaging system should allow pinning to a specific git hash. Yes the sha1 hash used by github has flaws, but it makes swapping out a legitimate package for a compromised one much more difficult.

**Comment by ksandvik** [comment]
This is just my own opinion, but having a 'standard' like pyproject.toml and let various groups experiment out the perfect packaging/dependency system would give more fruitful results than bake it into the compiler itself or having a std library with so many components, where the Python team has now realized that it was not so optimal after all. With meta-programming you might achieve similar results like Zig, so this could even be something code could have implemented rather than external tools nor the compiler itself.

**Reply by lsh** [reply]
A couple of notes on this:

1. This comment seems to conflate the concept of allowing using a URL as a package identifier (the proposal) and the actual build system (outside the scope).

2. Including a pyproject.toml specifically fails the use case I mentioned of _single file reproducible scripts_. I don't think installing all your package dependencies globally is a good solution, but requiring every script to be bundled into a whole project, and for users to download and set up a build system just for a script seems like an awful lot of work. It would dissuade me from using Mojo for scripts, and I'm sure many others. Though build systems are outside the scope of this proposal entirely, I mentioned the cargo example specifically because adding the scripting use case of listing dependencies later is much more difficult than starting with it.

3. A large portion of the Python ecosystem in general is unsatisfied with the pyproject.toml style competition going on. According to the [Python Packaging User Survey](https://drive.google.com/file/d/1U5d5SiXLVkzDpS0i1dJIA4Hu5Qg704T9/view) ran by the PIP maintainers:
    * Only 43% of respondents thought "Python packaging 'just works'".
    * 50% of respondents think packaging is "too complex" (with 22% neither agreeing or disagreeing).
    * and only 29% of respondents thought that the existence of multiple tools is beneficial for the python packaging ecosystem.
    * 75% of respondents preferred to have a clearly defined "official" workflow.

**Reply by drunkwcodes** [reply]
@lsh  I use `pdm` for my projects now. It's the most cutting-edge workflow in Python nowadays. It solves most of the problems you mentioned. It uses venv by default to isolate dependencies by project, and has a hashed lock file. It also manages and edits `pyproject.toml` for you.

`pdm` makes `pyproject.toml` a single reproducible configuration file.

The cutting-edge quality is not about percentages, you should use it. https://github.com/pdm-project/pdm

**Reply by mojodojodev** [reply]
@drunkwcodes this is the creator of node talking about his regrets using this style of implementation and why he created an entire new runtime in Deno to fix this regret : https://youtu.be/M3BM9TB-8yA?t=590

This proposal is mainly inspired by Deno, which is starting to eat into Pythons market share for simple scripting precisely because of this feature.

**Reply by lsh** [reply]
I will check out pdm, but it doesn't really answer the use case here, which is specifically about scripting files that can exist with pinned dependencies without a pyproject.toml

**Reply by drunkwcodes** [reply]
@lsh Github dependabot alert is annoying. It keeps telling me to change the `jQuery` version due to its CVE vulnerabilities.

If you pin the version number in the code, you will change the code when it happens. I would rather change the config files only than change the code.

Python can pin the version down by `__version__` for now, but it is not a fashionable way to write robust code.

**Reply by lsh** [reply]
I could be misinterpreting, but this feels like you're still talking about project scope as opposed to scripting scope.

**Reply by drunkwcodes** [reply]
@mojodojodev I watched that before. It looks inspiring now.

@frostming Deno is starting to eat into Pythons market share for simple scripting!

But `pdm` is my fav tool, would you use Mojo and port `pdm` to Mojo?

**Reply by mojodojodev** [reply]
@drunkwcodes I gave `pdm` a try, it is indeed very good, love the lack of venv and global caching, going to switch to this for personal projects :)

I still think `url` imports are the way to go to make scripting in a single file easier, and having less files to manage and things to learn, but `pdm` is a great template if they decide to go the PEP 517/621 route.

**Reply by andre1sk** [reply]
In Python "let various groups experiment out the perfect packaging/dependency system would give more fruitful results" never materialized in anything even remotely as usable as say what Rust has. In every instance this was tried it resulted in fragmented package ecosystem that was pain for the user.

**Comment by drunkwcodes** [comment]
OK, guys, it's happening.

[PEP 722: Dependency specification for single-file scripts](https://discuss.python.org/t/pep-722-dependency-specification-for-single-file-scripts/29905)

Please discuss the single-file requirements thing in the discourse.

**Reply by lsh** [reply]
Thanks for sharing this. Clearly relevant! Though Mojo still has to figure out its relationship to PIP

**Reply by drunkwcodes** [reply]
The PEP has been rewritten to stand on its own rather than build atop 722.

https://discuss.python.org/t/pep-723-embedding-pyproject-toml-in-single-file-scripts/31151

Put entire Toml file into `__pyproject__` is insane!

**Comment by ksandvik** [comment]
I've been following the hundreds on comments on both PIP discussions and I don't put personally much faith on either one being accepted. But let's see. Mojo that builds standalone binaries -- such as for tools -- will have different requirements for a package manager. Maybe same requirements as for rust/cargo or golang.

**Comment by martin-braun** [comment]
Let's not forget to mention that using URLs for package dependency management does not mean that there is no possibility to have a lock file available. Deno offers an argument to define a lock file, which will be created if it does not exist, otherwise it will pull dependencies using that lock file.

**Comment by DanteOz** [comment]
Should be noted that Deno is moving away from https import. 
https://youtu.be/5DX49vzLfqw?t=941

**Reply by lsh** [reply]
Before I dive in I want to flag that URL imports will continue to work; Deno is moving towards a centralized registry to aid in project scale workflows. I think it's also important to note what they're moving _towards syntactically_. Deno is creating a centralized package registry. From what I understand from the video, these are the two ways imports can now work, both relying on a new `deno` package specifier:

In a single file, a user can just have
```ts
import oak from 'deno:oak@1.2.3';
```

or, in a project context
```jsonc
// deno.jsonc
{
    "imports": {
        "oak": "deno:oak@1.2.3",
        "lodash": "npm:lodash@1.0.0"
    }
}
```
```js
// file.ts
import oak from 'oak';
import _ from 'lodash';
```
Kevin specifically says:
> You don't need a deno.json file in your project

Which is arguably what this whole proposal is about.

---

All of that said, I don't see any conflicts with the above in this proposal.
The equivalent using this proposal would be:
```
import "pip:lodash=1.0.0"
import "mip:oak=1.2.3"
```
where `mip` is a theoretical mojo package manager.

In a multi-file workflow, If mojo gets a theoretical `moproject.toml` which looks like
```
[dependencies]
loadsh = { registry="pip", version="1.0.0" }
oak = { registry="mip", version="1.2.3" }
```
then imports should work as usual:
```
import lodash
import oak
```

I would also like to once again highlight that one of the motivations for this proposal is to push off the need for a centralized registry. Right now even the syntax of the language is in flux, so any registry would be accruing projects which will, in all likelihood, be non-functional in the very near future.

**Comment by ivellapillil** [comment]
I am all for url imports. It really gives the feeling of easy to use for first timers. No ceremonies to worry about.

For security, we can think of allow-list/whitelist for domains, or better logging in case we load dependencies from other domains etc. But having the language itself flexible to load from a url has great benefit. 

For example, I work for a large organization with closed source principles. We would have to setup a repository separately if we use package management approach. If we allow URL imports, it will make cross organization code reuse with just a whitelist, without having to setup an artifact repository like for maven etc.

**Comment by david-ragazzi** [comment]
+1

But I think use things like:

 `from "header:foo.h" import Bar` 

...a bit inconsistent and will generate new and complex rules and exceptions to specify the path of a package. Maybe follow the well known URL convention of browsers for local and remote files sounds more regular and familiar to us (we don't need reinvent the wheel). Examples:

`from "file://foo.h" import Bar`
`from "https://repository.com/std@0.1.2" import Bar`
`from "file://C:\home\Desktop\foo.h" import Bar`
`from "file:///home/Desktop/foo.h" import Bar`

` file://` is an old known prefix to local URLs.

For versioned PIP packages, for example, a helper function like `pip_url` (or whatever) could receive the package specifications and translate them into to the proper URL:

`from pip_url("Foo", "<=", "1.1.2") import Bar`

Above `pip_url()` will return a string like `https://pypi.org/project/Foo/1.1.2/`

**Reply by lsh** [reply]
Before I comment on the header file part, I want to point out that the way C/C++ interoperability will work in Mojo isn't well defined yet, so it was more of a bikeshed comment. I'm not particularly sold on the `file://` schema to link to header files. The goal of this proposal isn't to replace the linker, and I can imagine direct file paths to header files would be brittle when linking to libraries installed from package managers. Also even in the example above, you run into the weirdness of Windows versus Unix file paths.

I'm also not sure how the `pip_url` function is meant to be any simpler. If problem is that a user has to learn something new, then both solutions have roughly the same barrier. If anything, requiring a helper function for importing a library seems a little daunting to me.

On a more meta level, I like the idea that I know what _kind_ of import I'm getting. With prefixes like `c:header.h` or `pip:requests`, I would know that I'm importing C or Python respectively, versus having to scan over the URL to figure out whether FFI is occurring or not. An alternative (that I consider a little too verbose) is to use the [package-url](https://github.com/package-url/purl-spec/) schema as proposed in [PEP-725](https://peps.python.org/pep-0725/).

**Reply by david-ragazzi** [reply]
Well... It's the way that browsers handle local and remote address.. It was just a suggestion to an existent solution to make uniform the things... And if we are talking about URLs, then... 

The helper function was to put all external item in `from` as a string and avoid internal analysis by the compiler of such string as it already is a full path to a file... 

It is a fine solution to me. But as you said, this is my bikesheding too..

> An alternative (that I consider a little too verbose) is to use the [package-url](https://github.com/package-url/purl-spec/) schema as proposed in [PEP-725](https://peps.python.org/pep-0725/).

Hmmm.... I didn't know that there was a standardization for external dependencies for Python.. Maybe it is a common solution that I agree...
